File: al2ka.py
"""Convert Anilist activities to human readable format for volume start date and end date"""
import json
import re
from datetime import datetime

import requests

def make_anilist_request(mediaId, page = 1):
    """Makes a request to AniList"""
    url = "https://graphql.anilist.co"
    query = """
    query ($mediaId: Int, $page: Int) {
        Page(page: $page, perPage: 500) {
                activities(userId: 5613718, mediaId: $mediaId) {
                ... on ListActivity {
                    createdAt
                    media {
                        title {
                            userPreferred
                        }
                    }
                    status
                    progress
                }
            }
        }
    }
    """
    variables = {
        "mediaId": mediaId,
        "page": page
    }

    return requests.post(url, json={"query": query, "variables": variables}, timeout=60)

def check_overlaps(a, b):
    """Checks for any overlaps between two arrays"""
    for i in a:
        if i in b:
            return True
    return False

def main():
    """Entry point"""
    volumes = []

    with open("json/HellParadise.json", "r", encoding="utf8") as file:
        contents = file.read()
        jayson = json.loads(contents)
        mediaid = jayson["id"]
        volumes = jayson["volumes"]

    response = make_anilist_request(mediaid)
    if not response.ok:
        print(f"[ERROR] AniList returned an unsuccessful status code: {response.status_code}")
        exit(1)

    json_dict = response.json()
    activities = json_dict["data"]["Page"]["activities"]

    results = []
    for i, volume in enumerate(volumes):
        volume_start = volume["start"]
        volume_end = volume["end"]
        started_reading = None
        finished_reading = None

        for activity in activities:
            created = activity["createdAt"]

            # If the volume is the last volume and the activity is completed
            if volume == volumes[-1] and activity["status"] == "completed":
                if started_reading is None:
                    started_reading = created
                elif created < started_reading:
                    started_reading = created

                if finished_reading is None:
                    finished_reading = created
                elif created > finished_reading:
                    finished_reading = created
                continue

            if activity["status"] != "read chapter":
                continue

            regex = re.findall(r"(\d+)", activity["progress"])
            if regex is None:
                continue

            my_start = int(regex[0])
            my_end = int(regex[-1])
            if check_overlaps(range(volume_start, volume_end + 1), range(my_start, my_end + 1)):
                if started_reading is None:
                    started_reading = created
                elif created < started_reading:
                    started_reading = created

                if finished_reading is None:
                    finished_reading = created
                elif created > finished_reading:
                    finished_reading = created

        results.append({"volume": i + 1, "started_reading": started_reading, "finished_reading": finished_reading})

    for r in results:
        volume = r["volume"]
        finished = r["finished_reading"]
        time = datetime.utcfromtimestamp(finished).strftime("%d/%m/%Y %H:%M")
        print(f"Volume {volume}")
        print(f"Finished: {time}")
        print("------------------------")

if __name__ == "__main__":
    main()

File: al2ka.rb
require "net/http"
require "json"
require "time"
require "debug"

def make_al_request(media_id, page: 1)
    uri = URI("https://graphql.anilist.co/")
    query = <<-EOF
    query ($mediaId: Int, $page: Int) {
        Page(page: $page, perPage: 500) {
                activities(userId: 5613718, mediaId: $mediaId) {
                ... on ListActivity {
                    createdAt
                    media {
                        title {
                            userPreferred
                        }
                    }
                    status
                    progress
                }
            }
        }
    }
    EOF
    variables = {
        mediaId: media_id,
        page:
    }
    data = {
        query:,
        variables:
    }
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true

    request = Net::HTTP::Post.new(uri.path, {'Content-Type' => 'application/json'})
    request.body = data.to_json

    http.request(request)
end

def check_overlaps(a, b)
    a.each do |i|
        if b.include?(i)
            return true
        end
    end
    false
end

def create_read_issues(kaboom_id, results)
    user = User.first
    comic = Comic.find(kaboom_id)
    ris = []
    comic.ordered_issues.each_with_index do |issue, index|
        result = results[index]
        next unless result
        read_at = result[:finished_reading]
        ris << ReadIssue.new(user:, issue:, read_at: )
    end
    ris
end

def main
    volumes = []

    kaboom_id = 56
    uri = URI("https://raw.githubusercontent.com/crxssed7/al3ka/main/json/Saiki.json")
    response = Net::HTTP.get_response(uri)
    return unless response.is_a? Net::HTTPSuccess

    json = JSON.parse(response.body)
    media_id = json["id"]
    volumes = json["volumes"]

    al_response = make_al_request(media_id)
    return unless al_response.is_a? Net::HTTPSuccess

    al_json = JSON.parse(al_response.body)
    activities = al_json["data"]["Page"]["activities"]

    results = []
    volumes.each_with_index do |volume, index|
        volume_start = volume["start"]
        volume_end = volume["end"]
        started_reading = nil
        finished_reading = nil

        activities.each do |activity|
            created = activity["createdAt"]

            if volume == volumes.last && activity["status"] == "completed"
                if started_reading == nil
                    started_reading = created
                elsif created < started_reading
                    started_reading = created
                end

                if finished_reading == nil
                    finished_reading = created
                elsif created > finished_reading
                    finished_reading = created
                end

                next
            end

            next if activity["status"] != "read chapter"

            regex = /(\d+)/
            match = activity["progress"].scan(regex).flatten
            next unless match.any?

            my_start = match[0].to_i
            my_end = match[-1].to_i
            
            if check_overlaps((volume_start..volume_end).to_a, (my_start..my_end).to_a)
                if started_reading == nil
                    started_reading = created
                elsif created < started_reading
                    started_reading = created
                end

                if finished_reading == nil
                    finished_reading = created
                elsif created > finished_reading
                    finished_reading = created
                end
            end
        end

        results << {volume: index + 1, started_reading: Time.at(started_reading).utc, finished_reading: Time.at(finished_reading).utc}
    end

    create_read_issues(kaboom_id, results)
end

def visual
    main.map { [_1.read_at, _1.issue.issue_number] }
end